qiita_db/artifact.py

```diff
@@ -1115,23 +1119,98 @@ def descendants_with_jobs(self):
-        if edges:
+            lineage = nx.DiGraph()
+            edges = set()
             nodes = {}
-            for jid, pid, cid in edges:
-                if jid not in nodes:
-                    nodes[jid] = ('job', qdb.processing_job.ProcessingJob(jid))
-                if pid not in nodes:
-                    nodes[pid] = ('artifact', qdb.artifact.Artifact(pid))
-                if cid not in nodes:
-                    nodes[cid] = ('artifact', qdb.artifact.Artifact(cid))
-
-                lineage.add_edge(nodes[pid], nodes[jid])
-                lineage.add_edge(nodes[jid], nodes[cid])
-        else:
-            lineage.add_node(('artifact', self))
+            if sql_edges:
+                for jid, pid, cid in sql_edges:
+                    if jid not in nodes:
+                        nodes[jid] = ('job',
+                                      qdb.processing_job.ProcessingJob(jid))
+                    if pid not in nodes:
+                        nodes[pid] = ('artifact', qdb.artifact.Artifact(pid))
+                    if cid not in nodes:
+                        nodes[cid] = ('artifact', qdb.artifact.Artifact(cid))
+                    edges.add((nodes[pid], nodes[jid]))
+                    edges.add((nodes[jid], nodes[cid]))
+            else:
+                nodes[self.id] = ('artifact', self)
+                lineage.add_node(nodes[self.id])
+
+            # The code above returns all the jobs that have been successfully
+            # executed. We need to add all the jobs that are in all the other
+            # status. Approach: Loop over all the artifacts and add all the
+            # jobs that have been attached to them.
+            visited = set()
+            queue = nodes.keys()
+            while queue:
+                current = queue.pop(0)
+                if current not in visited:
+                    visited.add(current)
+                    n_type, n_obj = nodes[current]
+                    if n_type == 'artifact':
+                        # Add all the jobs to the queue
+                        for job in n_obj.jobs():
+                            queue.append(job.id)
+                            if job.id not in nodes:
+                                nodes[job.id] = ('job', job)
+
+                    elif n_type == 'job':
+                        jstatus = n_obj.status
+                        # If the job is in success we don't need to do anything
+                        # else since it would've been added by the code above
+                        if jstatus != 'success':
+                            # Connect the job with his input artifacts, the
+                            # input artifacts may or may not exist yet, so we
+                            # need to check both the input_artifacts and the
+                            # pending properties
+                            for in_art in n_obj.input_artifacts:
+                                edge = (nodes[in_art.id], nodes[n_obj.id])
```

Comment:
These 3 lines are used over and over, perhaps add a _function for this?
```
edge = (nodes[in_node_id], nodes[n_obj.id])
if edge not in edges:
   edges.add(edge)
```

Reply:
Done